fluent/src/bundle.js

```diff
@@ -191,23 +191,88 @@ export default class FluentBundle {
    * @param   {Array}              errors
    * @returns {?string}
    */
-  format(message, args, errors) {
-    // optimize entities which are simple strings with no attributes
-    if (typeof message === "string") {
-      return this._transform(message);
+
+  format(path, args, errors) {
+
+    var parts = path.split(".");
+    var id = parts[0];
+    var message = this._messages.get(id);
+
+    if (!this._messages.has(id)) {
+      errors.push(`Message not found: "${id}"`);
+      return null;
     }
 
     // optimize entities with null values
-    if (message === null || message.value === null) {
+    if (parts.length === 1) {
+      // optimize entities which are simple strings with no attributes
+      if (typeof message === "string") {
+        return this._transform(message);
+      }
+      // optimize simple-string entities with attributes
+      if (typeof message.value === "string") {
+        return this._transform(message.value);
+      }
+
+      if (message.value === null) {
+        return null;
+      }
+
+      return resolve(this, args, message, errors);
+    }
+
+    if (parts.length === 2) {
+      if (message.attrs === null) {
+        errors.push(`Message has no attributes: "${id}"`);
+        return null;
+      }
+
+      let attr = message.attrs[parts[1]];
```

Comment:
To keep it consistent with the `parts.length === 1` branch above, perhaps you could add an `if (typeof attr === "string")` fast path? I'm not sure they're needed and I'll do some perf testing when I get a chance. For now it would be best to keep both branches as similar to each other as possible.

Reply:
I'm not sure where to place this fast path and which return it correlates to. Should this go in line 229?